getline.ts/src/blockchain.ts

```diff
@@ -0,0 +1,237 @@
+     * @returns Newly deployed contract.
+     */
+    deploy(name: string, ...params: Array<any>): Promise<Contract>;
+    /**
+     * Returns coinbase of the blockchain client.
+     * @returns Coinbase address.
+     */
+    coinbase(): Promise<Address>;
+}
+
+/**
+ * A contract deployed on a blockchain.
+ */
+export class Contract {
+    /**
+     * Address of the contract on the blockchain.
+     */
+    public address: Address
+    /**
+     * Blockchain on which the contract is located.
+     */
+    private blockchain: Blockchain
+    private instance: Web3.ContractInstance
+
+    constructor(blockchain: Blockchain, instance: Web3.ContractInstance) {
+        this.blockchain = blockchain;
+        this.instance = instance;
+        this.address = new Address(this.blockchain, instance.address);
+    }
+
+    /**
+     * Call a contract method.
+     * @param methodName Name of method to call.
+     * @param params Arguments to pass to method.
+     * @returns Result of call.
+     */
+    public async call(methodName: string, ...params: Array<any>): Promise<any> {
+        let method: any = this.instance[methodName];
+        return new Promise<any>((resolve, reject)=>{
+            let opts = {
+                gas: 1000000,
+            }
+            method(...params, opts, (err, object)=>{
+                if (err != undefined) {
+                    reject(err);
+                    return;
+                }
+                resolve(object);
+            });
+        });
+    }
+}
+
+/**
+ * Implementation of the Blockchain interface that uses the Getline metabackend
+ * and web3.
+ */
+export class GetlineBlockchain {
+    private metabackend: MetabackendClient;
+    private network: string;
+    private web3: Web3;
+    private contractDefinitions: Array<pb.Contract> | undefined;
+
+    constructor(metabackend: MetabackendClient, network: string) {
+        this.metabackend = metabackend;
+        this.network = network;
+
+        let provider = new Web3.providers.HttpProvider("http://localhost:8545")
+        if (typeof window !== 'undefined' && typeof window['web3'] !== 'undefined') {
+            console.log("getline.ts: using injected web3")
+            provider = window['web3'].currentProvider;
+        } else {
+            console.log("getline.ts: connecting to node running on localhost")
+        }
+        this.web3 = new Web3(provider);
+        if (this.web3.version.network != this.network) {
+            throw new Error("web3 is connected to wrong network")
+        }
+        this.web3.eth.defaultAccount = this.web3.eth.accounts[0];
+    }
+
+    /**
+     * Waits until a given Web3 contract is deployed on the blockchain.
+     *
+     * This is a hack that had to be written for when web3.ether.contract.new 
+     * fails to run the creation callback for the second time. When we upgrade
+     * to web3 1.0 or another library we should be able to get rid of this.
```

Comment:
Got a link to a reported error?

Reply:
There's no error, it just stops calling back (web3 calls back twice - once for when the transaction is broadcasted, once when it goes into a block, by subscribing internally to new block events - as far as I could tell with a debugger, that subscription stopped firing, too). I don't have any logs from then, but @balinskia did also get this. And it seems to have happened with no code changes.